Client/game_sa/CObjectGroupPhysicalPropertiesSA.cpp

```diff
@@ -0,0 +1,369 @@
+/*****************************************************************************
+ *
+ *  PROJECT:     Multi Theft Auto
+ *  LICENSE:     See LICENSE in the top level directory
+ *  FILE:        game_sa/CObjectGroupPhysicalPropertiesSA.cpp
+ *  PURPOSE:     Objects dynamic physical properties handler class
+ *
+ *  Multi Theft Auto is available from http://www.multitheftauto.com/
+ *
+ *****************************************************************************/
+
+#include "StdInc.h"
+#include "CObjectGroupPhysicalPropertiesSA.h"
+extern CGameSA* pGame;
+
+CObjectGroupPhysicalPropertiesSAInterface* pObjectInfo = (CObjectGroupPhysicalPropertiesSAInterface*)ARRAY_ObjectGroupsDynamicInfo;
+std::unordered_map<unsigned char, CObjectGroupPhysicalPropertiesSAInterface*> CObjectGroupPhysicalPropertiesSA::ms_OriginalGroupProperties;
+
+CObjectGroupPhysicalPropertiesSA::CObjectGroupPhysicalPropertiesSA() : m_pInterface(nullptr)
+{}
+
+CObjectGroupPhysicalPropertiesSA::CObjectGroupPhysicalPropertiesSA(unsigned char ucObjectGroup) : m_ucObjectGroup(ucObjectGroup)
+{
+    m_pInterface = &pObjectInfo[ucObjectGroup];
+    m_bModified = MapFind(ms_OriginalGroupProperties, ucObjectGroup);    
+}
+
+CObjectGroupPhysicalPropertiesSAInterface* CObjectGroupPhysicalPropertiesSA::GetInterface() const
+{
+    return m_pInterface;
+}
+
+void CObjectGroupPhysicalPropertiesSA::SetGroup(unsigned char ucObjectGroup)
+{
+    m_pInterface = &pObjectInfo[ucObjectGroup];
+    m_ucObjectGroup = ucObjectGroup;
+    m_bModified = MapFind(ms_OriginalGroupProperties, ucObjectGroup);
+}
+
+unsigned char CObjectGroupPhysicalPropertiesSA::GetGroup() const
+{
+    return m_ucObjectGroup;
+}
+
+bool CObjectGroupPhysicalPropertiesSA::IsValid() const
+{
+    return m_pInterface != nullptr;
+}
+
+void CObjectGroupPhysicalPropertiesSA::ChangeSafeguard()
+{
+    if (m_bModified)
+        return;
+
+    m_bModified = true;
+    // Make copy of original
+    if (!MapFind(ms_OriginalGroupProperties, m_ucObjectGroup))
+    {
+        auto pOriginalCopy = new CObjectGroupPhysicalPropertiesSAInterface;
```

Comment:
this should really be a `std::unique_ptr`, as it's rather hard to follow where the memory is freed.